packages/wrangler/src/api/startDevWorker/RemoteRuntimeController.ts

```diff
@@ -1,30 +1,303 @@
+				);
+			} else if (err instanceof UserError) {
+				logger.error(err.message);
+			}
+			// we want to log the error, but not end the process
+			// since it could recover after the developer fixes whatever's wrong
+			else if ((err as { code: string }).code !== "ABORT_ERR") {
+				logger.error("Error while creating remote dev session:", err);
+			} else {
+				throw err;
+			}
+		}
+	}
+
+	async #previewToken(
+		props: Parameters<typeof createRemoteWorkerInit>[0] &
+			Parameters<typeof getWorkerAccountAndContext>[0]
+	): Promise<CfPreviewToken | undefined> {
+		try {
+			const init = await createRemoteWorkerInit({
+				bundle: props.bundle,
+				modules: props.modules,
+				accountId: props.accountId,
+				name: props.name,
+				legacyEnv: props.legacyEnv,
+				env: props.env,
+				isWorkersSite: props.isWorkersSite,
+				assetPaths: props.assetPaths,
+				format: props.format,
+				bindings: props.bindings,
+				compatibilityDate: props.compatibilityDate,
+				compatibilityFlags: props.compatibilityFlags,
+				usageModel: props.usageModel,
+			});
+
+			const { workerAccount, workerContext } = await getWorkerAccountAndContext(
+				{
+					accountId: props.accountId,
+					env: props.env,
+					legacyEnv: props.legacyEnv,
+					host: props.host,
+					routes: props.routes,
+					sendMetrics: props.sendMetrics,
+				}
+			);
+			if (!this.#session) {
+				return;
+			}
+
+			const workerPreviewToken = await createWorkerPreview(
+				init,
+				workerAccount,
+				workerContext,
+				this.#session,
+				this.#abortController.signal
+			);
+
+			return workerPreviewToken;
+		} catch (err: unknown) {
+			assert(err && typeof err === "object");
+			// we want to log the error, but not end the process
+			// since it could recover after the developer fixes whatever's wrong
+			// instead of logging the raw API error to the user,
+			// give them friendly instructions
+			if ((err as unknown as { code: string }).code !== "ABORT_ERR") {
+				// code 10049 happens when the preview token expires
+				if ("code" in err && err.code === 10049) {
+					logger.log("Preview session expired, fetching a new one");
+
+					this.#session = await this.#previewSession(props);
+					return this.#previewToken(props);
+				} else if (
+					!handleUserFriendlyError(err as ParseError, props.accountId)
+				) {
+					logger.error("Error on remote worker:", err);
+				}
+			} else {
+				throw err;
+			}
+		}
+	}
+
+	async #onBundleComplete({ config, bundle }: BundleCompleteEvent, id: number) {
+		try {
+			const routes = config.triggers
+				?.filter(
+					(trigger): trigger is Extract<Trigger, { type: "route" }> =>
+						trigger.type === "route"
+				)
+				.map((trigger) => {
+					const { type: _, ...route } = trigger;
+					if (!("custom_domain" in route)) {
+						return route.pattern;
+					}
+					return route;
+				});
```

Comment:
I had this code (unpushed 🤦) on my branch:

```suggestion
				.map((trigger): Route => {
					const { type: _, ...route } = trigger;

					if (
						"custom_domain" in route ||
						"zone_id" in route ||
						"zone_name" in route
					) {
						return route;
					} else {
						return route.pattern; // SimpleRoute
					}
				});
```

Just want to ensure it's functionally identical..?